---
title: "Integration"
author: "Neal Liu"
date: "2024-02-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

load libraries

```{r}

library(CCA)
library(MOFA2)
library(ggplot2)
library(dplyr)
library(tidyr)
library(ggExtra)
library(cowplot)
library(SummarizedExperiment)
library(biomaRt)
library(MultiAssayExperiment)
library(WGCNA)
library(plotly)
library(viridis)
library(pheatmap)
library(cluster)

source('Helper_functions.R')

```

load data

```{r}
ov_data <- readRDS(file = 'Data/Normalized_transcriptome&methylation_TCGA_OV.rds')
brca_data <- readRDS(file = 'Data/TCGA_BRCA_Normalized_transcriptome&methylation.rds')

```

Trying CCA with methylation & transcriptome, selecting features with high variance (median absolute deviation > 1. 
for OV, this yields 5k genes for transcriptome and 6.9k for methylation. For BRCA, this yields 3.6k genes fro transcriptome and 4.2k sites for methylation.)

```{r}

ov_selected_genes <- apply(assay(ov_data[['normalized_transcriptome_raw_counts']],'RUV_K4_on_raw_counts'),1 ,FUN = mad) >= 1
ov_selected_methylations <- apply(assay(ov_data[['normalized_methylation_M_values']],'RUV_K12_on_M_values'),1 ,FUN = mad) >= 1

print(paste0(sum(ov_selected_genes),', ',sum(ov_selected_methylations)))

```

```{r}
ov_data_to_use <- lineup_samples(list(ov_data[['normalized_transcriptome_raw_counts']], ov_data[['normalized_methylation_M_values']]))

ov_metadata <- ov_data_to_use[[1]]@colData

ov_transcriptome_data_to_use <- t(assay(ov_data_to_use[[1]],'RUV_K4_on_raw_counts')[(ov_selected_genes),]) # %>% scale()
ov_methylation_data_to_use <- t(assay(ov_data_to_use[[2]],'RUV_K12_on_M_values')[(ov_selected_methylations),]) # %>% scale()


```


```{r}

ov_cancor <- stats::cancor(ov_transcriptome_data_to_use, ov_methylation_data_to_use)

component <- 1

ov_cca_component1_transcriptome <- ov_transcriptome_data_to_use[,rownames(ov_cancor$xcoef)] %*% ov_cancor$xcoef[,2]
ov_cca_component1_methylation <- ov_methylation_data_to_use[,rownames(ov_cancor$ycoef)] %*% ov_cancor$ycoef[,1]

ggplot(mapping = aes(x= ov_cca_component1_transcriptome, y= ov_cca_component1_methylation, color = ov_metadata$subtype)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x=paste0('transcriptome component ', component), 
       y=paste0('methylation component ', component),
       color = 'Subtype',
       title = paste0('Canonical correlation component ', component,' of ovarian cancer')) +
  theme(axis.line = element_line(colour = "grey83", linewidth=1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.05)

```




```{r}
brca_selected_genes <- apply(assay(brca_data[['normalized_transcriptome_raw_counts']],'RUV_K8_on_HTseq_counts'),1 ,FUN = mad) >= 2
brca_selected_methylations <- apply(assay(brca_data[['normalized_methylation_M_values']],'RUV_K10_on_M_values'),1 ,FUN = mad) >= 1.7
brca_transcriptome_data_to_use <- t(assay(brca_data[['normalized_transcriptome_raw_counts']],'RUV_K8_on_HTseq_counts')[brca_selected_genes,]) # %>% scale()
brca_methylation_data_to_use <- t(assay(brca_data[['normalized_methylation_M_values']],'RUV_K10_on_M_values')[brca_selected_methylations,]) # %>% scale()

print(paste0(sum(brca_selected_genes),', ',sum(brca_selected_methylations)))

```

```{r}

brca_cancor <- stats::cancor(x= brca_transcriptome_data_to_use,
                             y= brca_methylation_data_to_use)

component <- 1

brca_cca_component1_transcriptome <- brca_transcriptome_data_to_use[,rownames(brca_cancor$xcoef)] %*% brca_cancor$xcoef[,component]
brca_cca_component1_methylation <- brca_methylation_data_to_use[,rownames(brca_cancor$ycoef)] %*% brca_cancor$ycoef[,component]

ggplot(mapping = aes(x= brca_cca_component1_transcriptome, y= brca_cca_component1_methylation, color = brca_data[['normalized_transcriptome_raw_counts']]@colData$subtype)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x=paste0('transcriptome component ', component), 
       y=paste0('methylation component ', component),
       color = 'Subtype',
       title = paste0('Canonical correlation component ', component,' of breast cancer')) +
  theme(axis.line = element_line(colour = "grey83", linewidth=1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.05)

```

Trying MOFA

```{r}

ov_mofa <- MOFA2::create_mofa(list(transcriptome = t(ov_transcriptome_data_to_use), 
                                   methylation = t(ov_methylation_data_to_use)))
model_opts <- MOFA2::get_default_model_options(ov_mofa)
model_opts$num_factors <- 5
ov_mofa <- MOFA2::prepare_mofa(ov_mofa, model_options = model_opts)
ov_mofa <- MOFA2::run_mofa(ov_mofa, use_basilisk= T )

```



```{r, fig.width=15, fig.height= 6}

p1 <- MOFA2::plot_variance_explained(ov_mofa)
p2 <- MOFA2::plot_factor(ov_mofa, factors = c(1:5), color_by = ov_metadata$subtype)

gridExtra::grid.arrange(p1, p2, ncol=2)

```

```{r}

MOFA2::plot_factors(ov_mofa, factors = c(1,2), alpha = 0.7, color_by = ov_metadata$subtype) +
  theme_minimal() +
  theme(axis.line = element_line(colour = "grey83", linewidth = 1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.1)


```


```{r}
rownames(brca_transcriptome_data_to_use) <- substr(rownames(brca_transcriptome_data_to_use), 9, 16)
rownames(brca_methylation_data_to_use) <- substr(rownames(brca_methylation_data_to_use), 9, 16)
brca_metadata <- brca_data[[1]]@colData

print(paste0(sum(brca_selected_genes),', ',sum(brca_selected_methylations)))
```

```{r}
brca_mofa <- MOFA2::create_mofa(list(transcriptome = t(brca_transcriptome_data_to_use),
                                     methylation = t(brca_methylation_data_to_use)))
model_params <- MOFA2::get_default_model_options(brca_mofa)
model_params$num_factors <- 5
brca_mofa <- MOFA2::prepare_mofa(brca_mofa, model_options = model_params)
brca_mofa <- MOFA2::run_mofa(brca_mofa, use_basilisk = T)

```

```{r, fig.width=15, fig.height= 6}

p1 <- MOFA2::plot_variance_explained(brca_mofa)
p2 <- MOFA2::plot_factor(brca_mofa, factors = c(1:5), color_by = brca_metadata$subtype)

gridExtra::grid.arrange(p1, p2, ncol=2)

```

```{r, fig.width=12, fig.height=5}

p1 <- MOFA2::plot_factors(brca_mofa, factors= c(1,2), color_by = brca_metadata$subtype, alpha = 0.6) + 
  theme_minimal() +
  labs(title = 'MOFA factors for BRCA') +
  theme(axis.line = element_line(colour = "grey83", linewidth = 1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.1)

brca_cca_component1_transcriptome <- brca_transcriptome_data_to_use[,rownames(brca_cancor$xcoef)] %*% brca_cancor$xcoef[,1]
brca_cca_component1_methylation <- brca_methylation_data_to_use[,rownames(brca_cancor$ycoef)] %*% brca_cancor$ycoef[,2]

p2 <- ggplot(mapping = aes(x= brca_cca_component1_transcriptome, y= brca_cca_component1_methylation, color = brca_data[['normalized_transcriptome_raw_counts']]@colData$subtype)) +
  geom_point(alpha = 0.6) +
  theme_minimal() +
  labs(x=paste0('transcriptome component 1'), 
       y=paste0('methylation component 2'),
       color = 'Subtype',
       title = 'CCA components for BRCA') +
  theme(axis.line = element_line(colour = "grey83", linewidth=1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.05)

p1 + p2

```

```{r}
plot_PCA(t(brca_transcriptome_data_to_use),'brca transcriptome', brca_metadata$subtype,'subtypes' )



```



```{r}
plot_PCA(t(brca_methylation_data_to_use),'brca methylation', brca_metadata$subtype,'subtypes')

```

```{r}

View(umap(brca_transcriptome_data_to_use)$layout)
View(Rtsne(brca_transcriptome_data_to_use)$Y)
View(run_PCA(t(brca_transcriptome_data_to_use), pcs= 3)$u)
View(dist((run_PCA(t(brca_transcriptome_data_to_use), pcs= 3)$u)))

distance <- dist(Rtsne(brca_transcriptome_data_to_use)$Y)
summary(silhouette(as.numeric(as.factor(brca_metadata$subtype)), distance))

```



```{r}

compute_silhouette <- function(matrix, label_vector, run_dim_reduction = F, reduction_method = 'pca', num_pcs = 3){
  # taking a matrix of either raw data or dimensionally reduced data (PCA, umap or t-sne), calculate sihouette coefficient for each cluster.
  # if you input raw data, dimensionality reduction will be done for you.
  
  if(run_dim_reduction){
    if(!(reduction_method %in% c('pca','umap','tsne'))){stop("You inputted an invalid dimensionality reduction method. You can choose from 'pca', 'umap' or 'tsne'.")}
    if(reduction_method == 'pca'){matrix <- run_PCA(matrix, pcs = num_pcs)$u}
    if(reduction_method == 'umap'){matrix <- umap(t(matrix))$layout}
    if(reduction_method == 'tsne'){matrix <- Rtsne(t(matrix))$Y}
  }
  
  
  
  
}


```


















