---
title: "Integration"
author: "Neal Liu"
date: "2024-02-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

load libraries

```{r}

library(CCA)
library(MOFA2)
library(ggplot2)
library(dplyr)
library(tidyr)
library(ggExtra)
library(cowplot)
library(SummarizedExperiment)
library(biomaRt)
library(MultiAssayExperiment)
library(WGCNA)
library(plotly)
library(viridis)
library(pheatmap)
library(cluster)
library(mclust)

source('Helper_functions.R')

base::set.seed(42)

```

load data

```{r}
ov_data <- readRDS(file = 'Data/Normalized_transcriptome&methylation_TCGA_OV.rds')
brca_data <- readRDS(file = 'Data/TCGA_BRCA_Normalized_transcriptome&methylation.rds')

```

Trying CCA with methylation & transcriptome, selecting features with high variance (median absolute deviation > 1. 
for OV, this yields 5k genes for transcriptome and 6.9k for methylation. For BRCA, this yields 3.6k genes fro transcriptome and 4.2k sites for methylation.)

```{r}

ov_selected_genes <- apply(assay(ov_data[['normalized_transcriptome_raw_counts']],'RUV_K4_on_raw_counts'),1 ,FUN = mad) >= 1
ov_selected_methylations <- apply(assay(ov_data[['normalized_methylation_M_values']],'RUV_K12_on_M_values'),1 ,FUN = mad) >= 1

print(paste0(sum(ov_selected_genes),', ',sum(ov_selected_methylations)))

```

```{r}
ov_data_to_use <- lineup_samples(list(ov_data[['normalized_transcriptome_raw_counts']], ov_data[['normalized_methylation_M_values']]))

ov_metadata <- ov_data_to_use[[1]]@colData

ov_transcriptome_data_to_use <- t(assay(ov_data_to_use[[1]],'RUV_K4_on_raw_counts')[(ov_selected_genes),]) # %>% scale()
ov_methylation_data_to_use <- t(assay(ov_data_to_use[[2]],'RUV_K12_on_M_values')[(ov_selected_methylations),]) # %>% scale()


```


```{r}

ov_cancor <- stats::cancor(ov_transcriptome_data_to_use, ov_methylation_data_to_use)

component <- 1

ov_cca_component1_transcriptome <- ov_transcriptome_data_to_use[,rownames(ov_cancor$xcoef)] %*% ov_cancor$xcoef[,2]
ov_cca_component1_methylation <- ov_methylation_data_to_use[,rownames(ov_cancor$ycoef)] %*% ov_cancor$ycoef[,1]

ggplot(mapping = aes(x= ov_cca_component1_transcriptome, y= ov_cca_component1_methylation, color = ov_metadata$subtype)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x=paste0('transcriptome component ', component), 
       y=paste0('methylation component ', component),
       color = 'Subtype',
       title = paste0('Canonical correlation component ', component,' of ovarian cancer')) +
  theme(axis.line = element_line(colour = "grey83", linewidth=1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.05)

```




```{r}
brca_selected_genes <- apply(assay(brca_data[['normalized_transcriptome_raw_counts']],'RUV_K8_on_HTseq_counts'),1 ,FUN = mad) >= 2
brca_selected_methylations <- apply(assay(brca_data[['normalized_methylation_M_values']],'RUV_K10_on_M_values'),1 ,FUN = mad) >= 1.7
brca_transcriptome_data_to_use <- t(assay(brca_data[['normalized_transcriptome_raw_counts']],'RUV_K8_on_HTseq_counts')[brca_selected_genes,]) # %>% scale()
brca_methylation_data_to_use <- t(assay(brca_data[['normalized_methylation_M_values']],'RUV_K10_on_M_values')[brca_selected_methylations,]) # %>% scale()

print(paste0(sum(brca_selected_genes),', ',sum(brca_selected_methylations)))

```

```{r}

brca_cancor <- stats::cancor(x= brca_transcriptome_data_to_use,
                             y= brca_methylation_data_to_use)

component <- 1

brca_cca_component1_transcriptome <- brca_transcriptome_data_to_use[,rownames(brca_cancor$xcoef)] %*% brca_cancor$xcoef[,component]
brca_cca_component1_methylation <- brca_methylation_data_to_use[,rownames(brca_cancor$ycoef)] %*% brca_cancor$ycoef[,component]

ggplot(mapping = aes(x= brca_cca_component1_transcriptome, y= brca_cca_component1_methylation, color = brca_data[['normalized_transcriptome_raw_counts']]@colData$subtype)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x=paste0('transcriptome component ', component), 
       y=paste0('methylation component ', component),
       color = 'Subtype',
       title = paste0('Canonical correlation component ', component,' of breast cancer')) +
  theme(axis.line = element_line(colour = "grey83", linewidth=1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.05)

```

Trying MOFA

```{r}

ov_mofa <- MOFA2::create_mofa(list(transcriptome = t(ov_transcriptome_data_to_use), 
                                   methylation = t(ov_methylation_data_to_use)))
model_opts <- MOFA2::get_default_model_options(ov_mofa)
model_opts$num_factors <- 5
ov_mofa <- MOFA2::prepare_mofa(ov_mofa, model_options = model_opts)
ov_mofa <- MOFA2::run_mofa(ov_mofa, use_basilisk= T )

```



```{r, fig.width=15, fig.height= 6}

p1 <- MOFA2::plot_variance_explained(ov_mofa)
p2 <- MOFA2::plot_factor(ov_mofa, factors = c(1:5), color_by = ov_metadata$subtype)

gridExtra::grid.arrange(p1, p2, ncol=2)

```

```{r}

MOFA2::plot_factors(ov_mofa, factors = c(1,2), alpha = 0.7, color_by = ov_metadata$subtype) +
  theme_minimal() +
  theme(axis.line = element_line(colour = "grey83", linewidth = 1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.1)


```


```{r}
rownames(brca_transcriptome_data_to_use) <- substr(rownames(brca_transcriptome_data_to_use), 9, 16)
rownames(brca_methylation_data_to_use) <- substr(rownames(brca_methylation_data_to_use), 9, 16)
brca_metadata <- brca_data[[1]]@colData

print(paste0(sum(brca_selected_genes),', ',sum(brca_selected_methylations)))
```

```{r}
brca_mofa <- MOFA2::create_mofa(list(transcriptome = t(brca_transcriptome_data_to_use),
                                     methylation = t(brca_methylation_data_to_use)))
model_params <- MOFA2::get_default_model_options(brca_mofa)
model_params$num_factors <- 5
brca_mofa <- MOFA2::prepare_mofa(brca_mofa, model_options = model_params)
brca_mofa <- MOFA2::run_mofa(brca_mofa, use_basilisk = T)

```

```{r, fig.width=15, fig.height= 6}

p1 <- MOFA2::plot_variance_explained(brca_mofa)
p2 <- MOFA2::plot_factor(brca_mofa, factors = c(1:5), color_by = brca_metadata$subtype)

gridExtra::grid.arrange(p1, p2, ncol=2)

```

```{r, fig.width=12, fig.height=5}

p1 <- MOFA2::plot_factors(brca_mofa, factors= c(1,2), color_by = brca_metadata$subtype, alpha = 0.6) + 
  theme_minimal() +
  labs(title = 'MOFA factors for BRCA') +
  theme(axis.line = element_line(colour = "grey83", linewidth = 1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.1)

brca_cca_component1_transcriptome <- brca_transcriptome_data_to_use[,rownames(brca_cancor$xcoef)] %*% brca_cancor$xcoef[,1]
brca_cca_component1_methylation <- brca_methylation_data_to_use[,rownames(brca_cancor$ycoef)] %*% brca_cancor$ycoef[,2]

p2 <- ggplot(mapping = aes(x= brca_cca_component1_transcriptome, y= brca_cca_component1_methylation, color = brca_data[['normalized_transcriptome_raw_counts']]@colData$subtype)) +
  geom_point(alpha = 0.6) +
  theme_minimal() +
  labs(x=paste0('transcriptome component 1'), 
       y=paste0('methylation component 2'),
       color = 'Subtype',
       title = 'CCA components for BRCA') +
  theme(axis.line = element_line(colour = "grey83", linewidth=1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.05)

p1 + p2

```

```{r}
plot_PCA(t(brca_transcriptome_data_to_use),'brca transcriptome', brca_metadata$subtype,'subtypes' )



```



```{r}
plot_PCA(t(brca_methylation_data_to_use),'brca methylation', brca_metadata$subtype,'subtypes')

```

```{r}

View(umap(brca_transcriptome_data_to_use)$layout)
View(Rtsne(brca_transcriptome_data_to_use)$Y)
View(run_PCA(t(brca_transcriptome_data_to_use), pcs= 3)$u)
View(dist((run_PCA(t(brca_transcriptome_data_to_use), pcs= 3)$u)))

distance <- dist(Rtsne(brca_transcriptome_data_to_use)$Y)
a <- data.frame(labels = levels(factor(brca_metadata$subtype)),
                silhouette_score = summary(silhouette(as.numeric(as.factor(brca_metadata$subtype)), distance))$clus.avg.widths)

```


```{r}
brca_subtype_silhouette <- compute_silhouette(t(brca_transcriptome_data_to_use), brca_metadata$subtype, run_dim_reduction = 'pca', 
                   num_pcs = 2, result_format = 'per_cluster')

ggplot(data = brca_subtype_silhouette, aes(x= labels, y= silhouette_score, fill = labels))+
  geom_bar(stat='identity')+
  theme_minimal()+
  labs(title = 'Silhouette scores for filtered transcriptome')+
  theme(axis.line = element_line(colour = "grey88", linewidth=1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.25)

```

```{r}
compute_silhouette((brca_mofa@expectations$Z$group1[,1:2]), brca_metadata$subtype, result_format = 'plot')

```

```{r}
compute_silhouette(data.frame(transcriptome = brca_transcriptome_data_to_use[,rownames(brca_cancor$xcoef)] %*% brca_cancor$xcoef[,1],
                              methylation = brca_methylation_data_to_use[,rownames(brca_cancor$ycoef)] %*% brca_cancor$ycoef[,2]),
                   brca_metadata$subtype, result_format = 'plot')

```

```{r}
compute_silhouette(data.frame(transcriptome = brca_transcriptome_data_to_use[,rownames(brca_cancor$xcoef)] %*% brca_cancor$xcoef[,2],
                              methylation = brca_methylation_data_to_use[,rownames(brca_cancor$ycoef)] %*% brca_cancor$ycoef[,1]),
                   brca_metadata$subtype, result_format = 'plot')

```


```{r}
compute_ARI(brca_mofa@expectations$Z$group1[,1:2], brca_metadata$subtype, run_dim_reduction = NULL, 
                   num_pcs = 2)

```
```{r}

all_silhouette_scores <- parallel::mclapply(list(run_PCA(t(brca_transcriptome_data_to_use), pcs = 5)$u,
                                                 run_PCA(t(brca_methylation_data_to_use), pcs = 5)$u,
                                                 brca_mofa@expectations$Z$group1),
                                            function(x){compute_silhouette(x, brca_metadata$subtype, 
                                                                           result_format = 'per_cluster')})

all_silhouette_scores <- do.call(rbind, all_silhouette_scores)
all_silhouette_scores$algorithm <- rep(c('filtered transcriptome PCA', 'filtered methylation PCA', 'MOFA'), each = 6)

plot_boxplot_categorical(all_silhouette_scores$silhouette_score, 
            all_silhouette_scores$algorithm, 
            names = c('Silhouette scores per subtype', 'algorithm')) +
  theme(axis.text.x = element_text(size = 10,angle = 45,hjust = 1)) +
  theme(legend.position = 'None') +
  labs(title = 'Silhouette scores for components 1 to 5', x=NULL)

```


```{r}
compute_ARI(matrix = data.frame(transcriptome = brca_transcriptome_data_to_use[,rownames(brca_cancor$xcoef)] %*% brca_cancor$xcoef[,2],
                              methylation = brca_methylation_data_to_use[,rownames(brca_cancor$ycoef)] %*% brca_cancor$ycoef[,1]),
                   label_vector = brca_metadata$subtype,
            )

```

```{r}
devtools::install_github("immunogenomics/lisi", type = 'binary')
BiocManager::install("maftools")
library(lisi)

LISI_scores <- compute_lisi(run_PCA(t(brca_transcriptome_data_to_use)), 
                            meta_data = as.data.frame(do.call(cbind, brca_metadata@listData)), 
                            label_colnames = c('batch', 'subtype')) |> cbind.data.frame(dplyr::select(bm_data@meta.data, c('lane', 'donor', 'celltype.l2')))

```

```{r}
unique_methylations <- strsplit(brca_data[[2]]@rowRanges@elementMetadata$gene, ';') %>% unlist() %>% unique()
sum(!rownames(brca_data[[1]]) %in% unique_methylations)
sum(!unique_methylations %in% rownames(brca_data[[1]]))

rownames(brca_data[[1]])[!rownames(brca_data[[1]]) %in% unique_methylations]

hist(assay(brca_data[[2]], 2)[!brca_data[[2]]@rowRanges@elementMetadata$gene %in% rownames(brca_data[[1]]),],
     main = 'Methylations that are not in transcriptome')
```

```{r}
hist(assay(brca_data[[2]], 2)[brca_data[[2]]@rowRanges@elementMetadata$gene %in% rownames(brca_data[[1]]),],
     main = 'Methylations that are in transcriptome')
```

```{r}

hist(assay(brca_data[[2]], 2)[grepl(';', brca_data[[2]]@rowRanges@elementMetadata$gene),],
     main = 'Methylations that map to multiple genes')

```

```{r}
brca_selected_genes_with_purity <- apply(assay(brca_data[[3]],3),1 ,FUN = mad) >= 2
brca_selected_methylations_with_purity <- apply(assay(brca_data[[4]],4),1 ,FUN = mad) >= 1.75
brca_transcriptome_data_to_use_with_purity <- t(assay(brca_data[[3]],3)[brca_selected_genes_with_purity,])
brca_methylation_data_to_use_with_purity <- t(assay(brca_data[[4]],4)[brca_selected_methylations_with_purity,])

print(paste0(sum(brca_selected_genes_with_purity),', ',sum(brca_selected_methylations_with_purity)))

```

```{r}

rownames(brca_transcriptome_data_to_use_with_purity) <- substr(rownames(brca_transcriptome_data_to_use_with_purity), 9, 16)
rownames(brca_methylation_data_to_use_with_purity) <- substr(rownames(brca_methylation_data_to_use_with_purity), 9, 16)

brca_mofa_with_purity <- MOFA2::create_mofa(list(transcriptome = t(brca_transcriptome_data_to_use_with_purity),
                                     methylation = t(brca_methylation_data_to_use_with_purity)))
model_params <- MOFA2::get_default_model_options(brca_mofa_with_purity)
model_params$num_factors <- 5
brca_mofa_with_purity <- MOFA2::prepare_mofa(brca_mofa_with_purity, model_options = model_params)
brca_mofa_with_purity <- MOFA2::run_mofa(brca_mofa_with_purity, use_basilisk = T)

```

```{r, fig.width=15, fig.height= 6}

p1 <- MOFA2::plot_variance_explained(brca_mofa_with_purity)
p2 <- MOFA2::plot_factor(brca_mofa_with_purity, factors = c(1:5), color_by = brca_metadata$subtype)

gridExtra::grid.arrange(p1, p2, ncol=2)

```

```{r}
p1 <- MOFA2::plot_factors(brca_mofa_with_purity, factors= c(1,2), color_by = brca_metadata$subtype, alpha = 0.6) + 
  theme_minimal() +
  labs(title = 'MOFA factors for BRCA') +
  theme(axis.line = element_line(colour = "grey83", linewidth = 1.1),
          panel.border = element_rect(colour = "grey90", fill=NA, size=0.7),
          panel.grid.major = element_line(color = "grey96"),
          aspect.ratio = 1/1.1)

p1
```



```{r}
hist(colSums(assay(brca_data[[1]],1)))
hist(colSums(assay(brca_data[[1]],2)))

```



```{r}
devtools::install_github("linnykos/tiltedCCA")
library(tiltedCCA)



```

```{r}
a <- sort(abs(brca_mofa@expectations$W$transcriptome[,1]), decreasing = T)
b <- sort(abs(brca_mofa@expectations$W$methylation[,1]), decreasing = T)

names(a[1:30])
methylation_names <- strsplit(brca_data[[2]]@rowRanges@elementMetadata$gene[rownames(brca_data[[2]]) %in% names(b[1:30])], ';') %>% unlist()

base::intersect(names(a[1:30]), methylation_names)

```

```{r, fig.height= 6}
library(ggridges)

top_methylations_in_transcriptome <- methylation_names[methylation_names %in% rownames(assay(brca_data[[1]],4))]

plot_ridgeplot(t(assay(brca_data[[1]],4)[top_methylations_in_transcriptome,]), 
               top_methylations_in_transcriptome,
               title = 'Corresponding transcripts of the top methylations',
               aspect_ratio = 1/0.9)


```



```{r, fig.height= 7}
plot_ridgeplot(t(assay(brca_data[[2]],5)[which(rownames(assay(brca_data[[2]],5)) %in% names(b[1:30])), 
                                         which(brca_data[[2]]@colData$tissue_type == 'Normal')]),
               label_vector = brca_data[[2]]@rowRanges$gene[which(rownames(assay(brca_data[[2]],5)) %in% names(b[1:30]))],
               aspect_ratio = 1/0.6) +
  labs(y= ('Methylation gene mapping'), x=('M value'), title = 'Methylation M values of the top methylations')+
  coord_cartesian(xlim = c(-12,9))

```

```{r, fig.height= 7}
plot_ridgeplot(t(assay(brca_data[[2]],5)[which(rownames(assay(brca_data[[2]],5)) %in% names(b[1:30])), 
                                         which(brca_data[[2]]@colData$subtype == 'Basal')]),
               label_vector = brca_data[[2]]@rowRanges$gene[which(rownames(assay(brca_data[[2]],5)) %in% names(b[1:30]))],
               aspect_ratio = 1/0.6) +
  labs(y= ('Methylation gene mapping'), x=('M value'), title = 'Methylation M values of the top methylations')+
  coord_cartesian(xlim = c(-12,9))

```

```{r, fig.height= 7}
plot_ridgeplot(t(assay(brca_data[[2]],5)[which(rownames(assay(brca_data[[2]],5)) %in% names(b[1:30])), 
                                         which(brca_data[[2]]@colData$subtype == 'Basal')]),
               label_vector = brca_data[[2]]@rowRanges$gene[which(rownames(assay(brca_data[[2]],5)) %in% names(b[1:30]))],
               aspect_ratio = 1/0.6) +
  labs(y= ('Methylation gene mapping'), x=('M value'), title = 'Methylation M values of the top methylations')+
  coord_cartesian(xlim = c(-12,9))

```


```{r, fig.height= 7}
plot_ridgeplot(t(assay(brca_data[[2]],5)[which(rownames(assay(brca_data[[2]],5)) %in% names(b[1:30])),]),
               label_vector = brca_data[[2]]@rowRanges$gene[which(rownames(assay(brca_data[[2]],5)) %in% names(b[1:30]))],
               aspect_ratio = 1/0.6) +
  labs(y= ('Methylation gene mapping'), x=('M value'), title = 'Methylation M values of the top methylations')+
  coord_cartesian(xlim = c(-12,9))

```


```{r, fig.width = 8}
plot_GO_enrichment(brca_data[[1]]@elementMetadata$EnsemblGene_ids[which(brca_data[[1]]@elementMetadata$hgnc_symbol_BioMart %in% names(a[1:30]))],
                   'ENSEMBL',
)+ labs(title = 'GO enrichment of top 30 transcripts in MOFA compoennt 1')

```

```{r, fig.width= 8}
p <- plot_GO_enrichment(brca_data[[1]]@elementMetadata$EnsemblGene_ids[which(brca_data[[1]]@elementMetadata$hgnc_symbol_BioMart %in% methylation_names)],
                   'ENSEMBL',
)

p + labs(title = 'GO enrichment of top 30 methylations')
```



```{r, fig.height= 6}
a <- sort(abs(brca_mofa_with_purity@expectations$W$transcriptome[,1]), decreasing = T)
b <- sort(abs(brca_mofa_with_purity@expectations$W$methylation[,1]), decreasing = T)

methylation_names <- strsplit(brca_data[[3]]@elementMetadata$gene_name.[rownames(brca_data[[4]]) %in% names(b[1:30])], ';') %>% unlist()

top_methylations_in_transcriptome <- methylation_names[methylation_names %in% rownames(assay(brca_data[[3]],2))]

plot_ridgeplot(t(assay(brca_data[[3]],2)[top_methylations_in_transcriptome,]), 
               top_methylations_in_transcriptome,
               title = 'Corresponding transcripts of the top methylations',
               aspect_ratio = 1/0.9)


```

```{r}
plot_ridgeplot(t(assay(brca_data[[4]],4)[which(rownames(brca_data[[3]]) %in% names(b[1:30])),]),
               label_vector = brca_data[[2]]@rowRanges$gene[which(rownames(assay(brca_data[[3]],2)) %in% names(b[1:30]))],
               aspect_ratio = 1/0.6) +
  labs(y= ('Methylation gene mapping'), x=('M value'), title = 'Methylation M values of the top methylations')+
  coord_cartesian(xlim = c(-12,9))

```


```{r}

plot_PCA((assay(brca_data[[1]],4)[names(a[1:30]),]), 'top MOFA factor 1 transcripts',brca_metadata$subtype, 'subtype')

```

```{r}

plot_PCA((assay(brca_data[[2]],5)[names(b[1:30]),]), 'top MOFA factor 1 methylations',brca_metadata$subtype, 'subtype')



```

```{r}
plot_ridgeplot(assay(brca_data[[1]],4)['TMEM132C',], brca_metadata$subtype)


```

```{r}





```









