---
title: "TCGA_BRCA"
author: "Neal Liu"
date: "2024-01-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries

```{r}

library(TCGAbiolinks)
library(RUVIIIPRPS)
library(MOFA2)
library(ggplot2)
library(dplyr)
library(tidyr)
library(edgeR)
library(parallel)
library(readr)
library(ggExtra)
library(cowplot)
library(SummarizedExperiment)
library(biomaRt)
library(umap)
library(Rtsne)
library(singscore)
library(MultiAssayExperiment)
library(WGCNA)
library(plotly)
library(viridis)
library(ComplexHeatmap)
library(pheatmap)

source('Helper_functions.R')

library(devtools)

```


```{r}

data_transcriptome <- readRDS('Data/BRCA (Breast invasive carcinoma).rds')
data_methylation <- readRDS('Data/BRCA methylation illumina 27k.rds')
data_methylation_450k <- readRDS('Data/BRCA Methylation 450k.rds')


```

```{r}

sum(edgeR::filterByExpr(assay(data_transcriptome,'HTseq_counts'), min_count = 10, min.total.count = 80))

p1 <- plot_PCA(log2(assay(data_transcriptome, 'HTseq_FPKM')+1)[edgeR::filterByExpr(assay(data_transcriptome,'HTseq_counts'), min_count = 10, min.total.count = 80),],'fpkum uq transcriptome', data_transcriptome@colData$BatchId_mda, 'batch')

p2 <- plot_PCA(log2(assay(data_transcriptome, 'HTseq_FPKM')+1)[edgeR::filterByExpr(assay(data_transcriptome,'HTseq_counts'), min_count = 10, min.total.count = 80),],'fpkum uq transcriptome', data_transcriptome@colData$subtype_BRCA_Subtype_PAM50__RNAseq, 'subtype')

filtered_transcriptome <- data_transcriptome[edgeR::filterByExpr(assay(data_transcriptome,'HTseq_counts'), min_count = 10, min.total.count = 80),]

p1


```
```{r}
p2

```

```{r}

plot_3D_PCA(log2(assay(data_transcriptome, 'HTseq_FPKM')+1)[edgeR::filterByExpr(assay(data_transcriptome,'HTseq_counts'), min_count = 10, min.total.count = 80),], label_vector = data_transcriptome@colData$subtype_BRCA_Subtype_PAM50__RNAseq)

```


```{r, fig.width=19, fig.height= 6}
plot_RLE(log2(assay(data_transcriptome,'HTseq_counts')+1)[edgeR::filterByExpr(assay(data_transcriptome,'HTseq_counts'), min_count = 10, min.total.count = 80),1:500], data_transcriptome@colData$BatchId_mda[1:500], ylimit = c(-4, 4))

```

```{r}
data_methylation <- data_methylation[which(apply(assay(data_methylation), 1, function(x) {sum(is.na(x))/length(x)}) <= 0.2), ]
assay(data_methylation) <- replace_all_NAs_with_row_means(assay(data_methylation))
assay(data_methylation, 'M_values') <- log2(assay(data_methylation)/(1-assay(data_methylation)))

plot_PCA(assay(data_methylation,'M_values'),'M values', data_methylation@colData$paper_BRCA_Subtype_PAM50, 'pam50 subtype')

```

```{r, fig.width=19, fig.height= 6}
plot_RLE(assay(data_methylation,'M_values'),data_methylation@colData$site_of_resection_or_biopsy, ylimit = c(-4,4))


```

```{r}

# remove methylations that have more than 10% NAs and methylations that do not map to a gene. this ends up with around 330k rows.
data_methylation_450k <- data_methylation_450k[which(apply(assay(data_methylation_450k), 1, function(x) {sum(is.na(x))/length(x)}) <= 0.1), ]
data_methylation_450k <- data_methylation_450k[!is.na(data_methylation_450k@rowRanges@elementMetadata@listData$gene),]

# replace remaining NAs with row means and calculate M values

# The replace NAs calculation takes a while so we're going to save the object and load it
# assay(data_methylation_450k) <- replace_all_NAs_with_row_means(assay(data_methylation_450k))

data_methylation_450k <- readRDS('Data/Computationally intensive objects/BRCA 450k processed.rds')

assay(data_methylation_450k, 'M_values') <- log2(assay(data_methylation_450k)/(1-assay(data_methylation_450k)))

# methylation_450k_pca <- list(BiocSingular::runSVD(t(assay(data_methylation_450k,'M_values')),k=10,center=T))[[1]]
methylation_450k_pca <- readRDS('Data/Computationally intensive objects/BRCA 450k pca.rds')

plot_PCA(methylation_450k_pca,'M values', data_methylation_450k@colData$paper_BRCA_Subtype_PAM50, 'pam50 subtype', is_pca_obj = T, pcs = c(1,2))

```

Since the methylations are split into 27k and 450k, both having a significant amount of patients, 
it might be a good idea to combine them by keeping common sites between both. However, we must first 
check if there is significant effects between the two platforms because we don't know if they use 
different reagents etc which will have effects on the data distribution. 


```{r}

common_methylations <- intersect(rownames(assay(data_methylation,'M_values')),rownames(assay(data_methylation_450k,'M_values')))
common_patients <- intersect(data_methylation_450k@colData$patient, data_methylation@colData$patient)

`27k_test_matrix` <- assay(data_methylation,'M_values')[common_methylations, sort(which(data_methylation@colData$patient %in% common_patients))]
`450k_test_matrix` <- assay(data_methylation_450k,'M_values')[common_methylations, sort(which(data_methylation_450k@colData$patient %in% common_patients))]

methylation_platform_correlations <- as.matrix(unlist(lapply(1:length(common_methylations), 
                                                             function(x) {stats::cor(`27k_test_matrix`[x,],
                                                                              `450k_test_matrix`[x,], 
                                                                              method='pearson')})))

p1 <- ggplot(mapping = aes(x= `27k_test_matrix`[,1], y= `450k_test_matrix`[,1])) +
  geom_point( alpha = 0.05) +
  theme_minimal() +
  labs(x = '27k (vial A)', y = '450k (vial B)')+
  ggtitle('M values of patient A0DC')+
  theme(panel.border=element_rect(colour = "grey80", fill=NA, size=0.8),
          aspect.ratio = 1/1.05,
          axis.line = element_line(colour = "grey75", linewidth = 1.1),
          panel.grid.major = element_line(color = "grey96"),)

# 0.9489513 correlation

p2 <- ggplot(mapping = aes(x= assay(data_methylation,1)[common_methylations, "TCGA-A7-A0DC-01A-11D-A00Y-05"], 
                     y= assay(data_methylation_450k,1)[common_methylations, "TCGA-A7-A0DC-01B-04D-A22R-05"])) +
  geom_point( alpha = 0.05) +
  theme_minimal() +
  labs(x = '27k (vial A)', y = '450k (vial B)')+
  ggtitle('Beta values')+
  theme(panel.border=element_rect(colour = "grey80", fill=NA, size=0.8),
          aspect.ratio = 1/1.05,
          axis.line = element_line(colour = "grey75", linewidth = 1.1),
          panel.grid.major = element_line(color = "grey96"),)

# 0.9682381 correlation 

gridExtra::grid.arrange(p1,p2, ncol=2)

```

```{r}

p1 <- ggplot(mapping = aes(x= assay(data_methylation,2)[common_methylations, "TCGA-BH-A18V-01A-11D-A12E-05"], 
                     y= assay(data_methylation_450k,2)[common_methylations, "TCGA-BH-A18V-06A-11D-A212-05"])) +
  geom_point( alpha = 0.05) +
  theme_minimal() +
  labs(x = '27k (primary tumour)', y = '450k (metastatic)')+
  ggtitle('M values of patient A18V')+
  theme(panel.border=element_rect(colour = "grey80", fill=NA, size=0.8),
          aspect.ratio = 1/1.05,
          axis.line = element_line(colour = "grey75", linewidth = 1.1),
          panel.grid.major = element_line(color = "grey96"),)

# 0.9138258 correlation

p2 <- ggplot(mapping = aes(x= assay(data_methylation,1)[common_methylations, "TCGA-BH-A18V-01A-11D-A12E-05"], 
                     y= assay(data_methylation_450k,1)[common_methylations, "TCGA-BH-A18V-06A-11D-A212-05"])) +
  geom_point( alpha = 0.05) +
  theme_minimal() +
  labs(x = '27k (primary tumour)', y = '450k (metastatic)')+
  ggtitle('Beta values')+
  theme(panel.border=element_rect(colour = "grey80", fill=NA, size=0.8),
          aspect.ratio = 1/1.05,
          axis.line = element_line(colour = "grey75", linewidth = 1.1),
          panel.grid.major = element_line(color = "grey96"),)

# 0.9146544 correlation 

gridExtra::grid.arrange(p1,p2, ncol=2)

```

```{r}

data_methylation_450k@colData <- data_methylation_450k@colData[colnames(data_methylation_450k@colData) %in% colnames(data_methylation@colData)]
data_methylation_450k@colData$platform <- '450k'


data_methylation@colData <- data_methylation@colData[colnames(data_methylation_450k@colData) %in% colnames(data_methylation@colData)]
data_methylation@colData$platform <- '27k'

illumina_27k_rowdata <-  data_methylation@rowRanges
# rownames(illumina_27k_rowdata) <- data_methylation@NAMES
illumina_450k_rowdata <-  data_methylation_450k@rowRanges
# rownames(illumina_450k_rowdata) <- data_methylation_450k@NAMES

```
  

```{r}
data_methylation_complete <- SummarizedExperiment(assays = list( beta_values  = cbind(assay(data_methylation_450k[common_methylations,],1), 
                                                                                      assay(data_methylation[common_methylations,],1)),
                                                                 M_values = cbind(assay(data_methylation_450k[common_methylations,],2), 
                                                                                      assay(data_methylation[common_methylations,],2))),
                                                  rowData = illumina_450k_rowdata[common_methylations,],
                                                  colData =  rbind(data_methylation_450k@colData, data_methylation@colData),
                                                  )

```


```{r}

all_methylation_pca <- run_PCA(assay(data_methylation_complete,'M_values'))

plot_PCA(all_methylation_pca,'M values of all methylation patients', data_methylation_complete@colData$platform, 'platform', is_pca_obj = T)


```

```{r}
plot_PCA(all_methylation_pca,'M values of all methylation patients', data_methylation_complete@colData$paper_BRCA_Subtype_PAM50 ,'subtypes', is_pca_obj = T)

```

```{r}
plot_3D_PCA(all_methylation_pca, data_methylation_complete@colData$paper_BRCA_Subtype_PAM50, is_PCA_obj = T)

```

```{r}
filtered_transcriptome@colData <- t(remove_NA_rows(t(data_transcriptome@colData)))

filtered_transcriptome_matched_samples <- filtered_transcriptome[, which(filtered_transcriptome@colData$sample__RNAseq %in% intersect(data_methylation_complete@colData$sample, filtered_transcriptome@colData$sample__RNAseq))]
data_methylation_complete_matched_samples <- data_methylation_complete[, which(data_methylation_complete@colData$sample %in% intersect(data_methylation_complete@colData$sample, filtered_transcriptome@colData$sample__RNAseq))]

```

```{r, fig.width=19, fig.height= 7}
plot_RLE(assay(data_methylation_complete_matched_samples,'M_values')[, 700:1000], data_methylation_complete_matched_samples@colData$platform[700:1000], ylimit = c(-5,5))

```
```{r}
data_methylation_complete_matched_samples@colData$Plate <- substr(data_methylation_complete_matched_samples@colData$barcode, 22,25)


```

```{r, fig.width=19, fig.height= 7}
plot_RLE((assay(data_methylation_complete_matched_samples,'M_values')[,which(data_methylation_complete_matched_samples@colData$platform == '450k')]), (data_methylation_complete_matched_samples@colData$Plate[which(data_methylation_complete_matched_samples@colData$platform == '450k')]), ylimit = c(-4,4))

```


```{r}
table(data_methylation_complete_matched_samples@colData$Plate)
table(filtered_transcriptome_matched_samples@colData$BatchId_mda)

data_methylation_complete_matched_samples <- data_methylation_complete_matched_samples[,!duplicated(data_methylation_complete_matched_samples@colData$sample)]
data_methylation_complete_matched_samples<- data_methylation_complete_matched_samples[,match(filtered_transcriptome_matched_samples@colData$sample__RNAseq, data_methylation_complete_matched_samples@colData$sample)]
data_methylation_complete_matched_samples@colData$singscore_purity <- filtered_transcriptome_matched_samples@colData$Purity_singscore

```


```{r, fig.width= 9}

methylation_normalized <- data_methylation_complete_matched_samples
methylation_normalized@colData$Plate <- factor(methylation_normalized@colData$Plate)
methylation_normalized@colData$platform <- factor(methylation_normalized@colData$platform)
methylation_normalized@colData$subtype <- filtered_transcriptome_matched_samples@colData$subtype_BRCA_Subtype_PAM50__RNAseq
methylation_normalized@colData$subtype[which(methylation_normalized@colData$definition == 'Solid Tissue Normal')] <- 'Normal Tissue'
methylation_normalized@colData$subtype[which(methylation_normalized@colData$subtype == 'Normal')] <- 'Normal like'


methylation_normalized <- RUVIIIPRPS::supervisedPRPS(methylation_normalized[,which(!is.na(methylation_normalized@colData$subtype))],                                                      'M_values',
                                                     bio.variable = 'subtype',
                                                     uv.variables = c('platform','Plate','singscore_purity'), 
                                                     batch.variable = 'Plate', 
                                                     min.sample.for.prps = 2,
                                                     min.sample.per.batch = 3,
                                                     apply.log = F,
                                                     pseudo.count = 1,
                                                     assess.se.obj = F,
                                                     assess.cor.variables = F,
                                                     save.se.obj = TRUE,
                                                     verbose = T)


# plotting prps map

prps_map <- as.data.frame(methylation_normalized@colData)[,c('Plate','subtype')] %>%
    dplyr::count(Plate, subtype)

prps_map$use <- ifelse(prps_map$n > 1, 'yes','no')

ggplot(prps_map, aes(x = Plate, y = subtype)) +
  geom_count(aes(color = use)) +
  geom_text(aes(
    label = n,
    hjust = 0.5,
    vjust = 0.5
  )) +
  xlab('Batch') +
  ylab('Confident subtypes') +
  theme_bw() +
  theme(axis.line = element_line(colour = 'grey35', size = 1.1),
    axis.title.x = element_text(size = 18),
    axis.title.y = element_text(size = 0),
    axis.text.x = element_text(size = 10,angle = 45,hjust = 1),
    axis.text.y = element_text(size = 12, angle = 45, hjust = 1),
    legend.position = 'none',
    panel.grid.major = element_line(color = "grey95"),
    aspect.ratio = 1/2)

```

```{r}
methylation_normalized <- RUVIIIPRPS::supervisedFindNCG(methylation_normalized, 'M_values',
                                                     bio.variable = 'subtype',
                                                     uv.variables = c('platform','Plate','singscore_purity'), 
                                                     no.ncg = 1000,
                                                     apply.log = F,
                                                     assess.se.obj = F)


```

```{r}
NCGs <- methylation_normalized@metadata$NCG

pca_NCGs <- run_PCA(assay(methylation_normalized,2)[NCGs, ])

plot_3D_PCA(pca_NCGs, methylation_normalized@colData$platform, is_PCA_obj = T)


```

```{r}
plot_3D_PCA(pca_NCGs, methylation_normalized@colData$Plate, is_PCA_obj = T)

```

```{r}

ggplot(mapping = aes(x=methylation_normalized@colData$singscore_purity, y=(pca_NCGs$u[,1])))+
  geom_point(alpha=0.62, color= 'grey30') + # alpha for transparency to see overlap
  labs(y="PC1 of NCGs", 
       x="Purity") +
  theme_minimal()+
  ggtitle('1st PC of negative control genes vs purity') +
  theme(panel.border=element_rect(colour = "grey80", fill=NA, size=0.8),
        aspect.ratio = 1/1.05,
        axis.line = element_line(colour = "grey75", linewidth = 1.1),
        panel.grid.major = element_line(color = "grey96"),)


```

```{r}
replicate.data=t(do.call(cbind, methylation_normalized@metadata[['PRPS']][['supervised']]))

methylation_normalized <- RUVIIIPRPS::ruvIIIMultipleK(se.obj = methylation_normalized, 
                                                      assay.name = 'M_values',
                                                      apply.log = F,
                                                      k = c(6,10,14),
                                                      ctl= methylation_normalized@metadata[['NCG']],
                                                      assess.se.obj = F, 
                                                      replicate.data = replicate.data,
                                                      save.se.obj = T,
                                                      remove.na = 'measurements')


```


```{r, fig.width= 12}

# normalized_methylation_pcas <- lapply(assays(methylation_normalized)[-1], run_PCA)
# the pca takes a while so we saved it.
normalized_methylation_pcas <- readRDS(file = 'Data/Computationally intensive objects/BRCA normalized methylation pcas.rds')

p1 <- plot_vector_correlation(normalized_methylation_pcas, c('M values', 'RUV k6', 'RUV k10', 'RUV k14'), methylation_normalized@colData$platform, 'Platform')

p2 <- plot_vector_correlation(normalized_methylation_pcas, c('M values', 'RUV k6', 'RUV k10', 'RUV k14'), methylation_normalized@colData$Plate, 'plate')

p3 <- plot_linear_correlation(normalized_methylation_pcas, c('M values', 'RUV k6', 'RUV k10', 'RUV k14'), methylation_normalized@colData$singscore_purity, 'purity')

gridExtra::grid.arrange(p1,p2, p3, ncol = 2, nrow = 2)

```

```{r,fig.width=19, fig.height= 7}

plot_RLE(assay(methylation_normalized,'RUV_K14_on_M_values')[,1:600], methylation_normalized@colData$Plate[1:600], ylimit = c(-5,5))


```

```{r,fig.width=19, fig.height= 7}

plot_RLE(assay(methylation_normalized,'RUV_K14_on_M_values')[,600:1168], methylation_normalized@colData$Plate[600:1168], ylimit = c(-5,5))


```
```{r}
p1 <- plot_PCA(normalized_methylation_pcas[[4]], 'RUV-K14', methylation_normalized@colData$subtype, 'subtypes', is_pca_obj = T)
p2 <- plot_PCA(normalized_methylation_pcas[[4]], 'RUV-K14', methylation_normalized@colData$platform, 'platform', is_pca_obj = T)

p2

```


```{r}
plot_3D_PCA(normalized_methylation_pcas[[4]], methylation_normalized@colData$subtype, is_PCA_obj = T)

```

Normalizing transcriptome raw counts, removing effects of: purity, library size and batch


```{r,fig.width= 9}

transcriptome_normalized <- filtered_transcriptome_matched_samples
transcriptome_normalized@colData$subtype <- transcriptome_normalized@colData$subtype_BRCA_Subtype_PAM50__RNAseq
transcriptome_normalized@colData$subtype[which(transcriptome_normalized@colData$definition__RNAseq == 'Solid Tissue Normal')] <- 'Normal Tissue'
transcriptome_normalized@colData$subtype[which(transcriptome_normalized@colData$subtype == 'Normal')] <- 'Normal like'
transcriptome_normalized@colData$batch <-transcriptome_normalized@colData$BatchId_mda
transcriptome_normalized@colData$log2_lib_size <- log2(apply(assay(transcriptome_normalized,'HTseq_counts'), 2, sum)+1)


transcriptome_normalized <- RUVIIIPRPS::supervisedPRPS(transcriptome_normalized[,which(!is.na(transcriptome_normalized@colData$subtype))],                                                      'HTseq_counts',
                                                     bio.variable = 'subtype',
                                                     uv.variables = c('log2_lib_size','Purity_singscore','batch'), 
                                                     batch.variable = 'batch', 
                                                     min.sample.for.prps = 2,
                                                     min.sample.per.batch = 3,
                                                     apply.log = T,
                                                     pseudo.count = 1,
                                                     assess.se.obj = F,
                                                     assess.cor.variables = F,
                                                     save.se.obj = TRUE,
                                                     verbose = T)


# plotting prps map

prps_map <- as.data.frame(transcriptome_normalized@colData)[,c('batch','subtype')] %>%
    dplyr::count(batch, subtype)

prps_map$use <- ifelse(prps_map$n > 1, 'yes','no')

ggplot(prps_map, aes(x = batch, y = subtype)) +
  geom_count(aes(color = use)) +
  geom_text(aes(
    label = n,
    hjust = 0.5,
    vjust = 0.5
  )) +
  xlab('Batch') +
  ylab('Confident subtypes') +
  theme_bw() +
  theme(axis.line = element_line(colour = 'grey35', size = 1.1),
    axis.title.x = element_text(size = 18),
    axis.title.y = element_text(size = 0),
    axis.text.x = element_text(size = 10,angle = 45,hjust = 1),
    axis.text.y = element_text(size = 12, angle = 45, hjust = 1),
    legend.position = 'none',
    panel.grid.major = element_line(color = "grey95"),
    aspect.ratio = 1/2)


```


```{r, fig.width=8}
transcriptome_normalized1 <- RUVIIIPRPS::supervisedFindNCG(transcriptome_normalized[,-which(transcriptome_normalized@colData$batch %in% names(table(transcriptome_normalized@colData$batch)[table(transcriptome_normalized@colData$batch) <= 3]))], 'HTseq_counts',
                                                     bio.variable = 'subtype',
                                                     uv.variables = c('log2_lib_size','Purity_singscore','batch'), 
                                                     no.ncg = 1000,
                                                     apply.log = T,
                                                     assess.se.obj = F,
                                                     )


NCGs <- transcriptome_normalized1@metadata$NCG
pca_NCGs <- run_PCA(assay(transcriptome_normalized1,1)[NCGs, ])

p1 <- ggplot(mapping = aes(x=transcriptome_normalized1@colData$Purity_singscore, y=(pca_NCGs$u[,1])))+
  geom_point(alpha=0.62, color= 'grey30') + # alpha for transparency to see overlap
  labs(y="PC1 of NCGs", 
       x="Purity") +
  theme_minimal()+
  ggtitle('1st PC of negative control genes vs purity') +
  theme(panel.border=element_rect(colour = "grey80", fill=NA, size=0.8),
        aspect.ratio = 1/1.05,
        axis.line = element_line(colour = "grey75", linewidth = 1.1),
        panel.grid.major = element_line(color = "grey96"),)

p2 <- ggplot(mapping = aes(x=transcriptome_normalized1@colData$log2_lib_size, y=(pca_NCGs$u[,1])))+
  geom_point(alpha=0.62, color= 'grey30') + # alpha for transparency to see overlap
  labs(y="PC1 of NCGs", 
       x="log2 libsize") +
  theme_minimal()+
  ggtitle('1st PC of negative control genes vs libsize') +
  theme(panel.border=element_rect(colour = "grey80", fill=NA, size=0.8),
        aspect.ratio = 1/1.05,
        axis.line = element_line(colour = "grey75", linewidth = 1.1),
        panel.grid.major = element_line(color = "grey96"),)


gridExtra::grid.arrange(p1,p2, ncol = 2)

```

```{r}
plot_3D_PCA(pca_NCGs, transcriptome_normalized1@colData$batch, is_PCA_obj = T)

```


```{r}
replicate.data=t(do.call(cbind, transcriptome_normalized1@metadata[['PRPS']][['supervised']]))

transcriptome_normalized <- RUVIIIPRPS::ruvIIIMultipleK(se.obj = transcriptome_normalized, 
                                                      assay.name = 'HTseq_counts',
                                                      apply.log = T,
                                                      k = c(4,6,8,10),
                                                      ctl= transcriptome_normalized1@metadata[['NCG']],
                                                      assess.se.obj = F, 
                                                      replicate.data = replicate.data,
                                                      save.se.obj = T,
                                                      remove.na = 'measurements')


```

```{r, fig.width= 13}

# normalized_transcriptome_pcas <- lapply(assays(transcriptome_normalized), run_PCA)

# The object is saved because it takes a while to compute. 
normalized_transcriptome_pcas <- readRDS(file= 'Data/Computationally intensive objects/BRCA normalized transcriptome pcas.rds')

p1 <- plot_vector_correlation(normalized_transcriptome_pcas, names(assays(transcriptome_normalized)), transcriptome_normalized@colData$batch, 'batch')

p2 <- plot_linear_correlation(normalized_transcriptome_pcas, names(assays(transcriptome_normalized)), transcriptome_normalized@colData$Purity_singscore, 'purity')

p3 <- plot_vector_correlation(normalized_transcriptome_pcas, names(assays(transcriptome_normalized)), transcriptome_normalized@colData$subtype, 'subtypes')

p4 <- plot_linear_correlation(normalized_transcriptome_pcas, names(assays(transcriptome_normalized)), transcriptome_normalized@colData$log2_lib_size, 'libsize')

gridExtra::grid.arrange(p1,p2, p3, p4, ncol = 2, nrow = 2)


```

```{r}

plot_3D_PCA(normalized_transcriptome_pcas[['RUV_K10_on_HTseq_counts']], transcriptome_normalized@colData$subtype, is_PCA_obj = T)


```

```{r, fig.width=11, fig.height=6}
p1 <- plot_PCA(normalized_transcriptome_pcas[['RUV_K10_on_HTseq_counts']],'RUV K10', transcriptome_normalized@colData$subtype, 'subtypes',is_pca_obj = T)

p2 <- plot_PCA(normalized_transcriptome_pcas[['RUV_K10_on_HTseq_counts']],'RUV K10', transcriptome_normalized@colData$subtype, 'subtypes',is_pca_obj = T, pcs = c(1,3))

gridExtra::grid.arrange(p1,p2, ncol=2)

```

```{r, fig.width= 11, fig.height=6}

p1 <- plot_PCA(pca_NCGs, 'transcriptome ncgs', transcriptome_normalized1@colData$log2_lib_size, 'log2 libsize', is_pca_obj = T, is_continuous = T, pcs= c(1,2))
p2 <- plot_PCA(assay(transcriptome_normalized, "RUV_K6_on_HTseq_counts")[NCGs,],'normalized ncgs', 
               transcriptome_normalized@colData$log2_lib_size, 'log2 libsize', is_pca_obj = F, is_continuous = T, pcs= c(1,2))

gridExtra::grid.arrange(p1,p2, ncol=2)

```


```{r, fig.width= 18, fig.height= 7}
plot_RLE(log2(assay(transcriptome_normalized,'HTseq_counts')[,1:600]+1), transcriptome_normalized@colData$batch[1:600], ylimit = c(-4,4))


```

```{r, fig.width= 12, fig.height= 6}
plot_RLE(log2(1+assay(transcriptome_normalized,'HTseq_counts'))[,1:300],transcriptome_normalized@colData$log2_lib_size[1:300], is_continuous = T, ylimit= c(-4,4), variable_name ='log2 libsize')
```

```{r, fig.width= 18, fig.height= 7}
plot_RLE(log2(assay(transcriptome_normalized,'HTseq_counts')[,600:1168]+1), transcriptome_normalized@colData$batch[600:1168], ylimit = c(-4,4))


```

```{r, fig.width= 18, fig.height= 7}
plot_RLE(assay(transcriptome_normalized,'RUV_K10_on_HTseq_counts')[,1:600], transcriptome_normalized@colData$batch[1:600], ylimit = c(-4,4))


```

```{r, fig.width= 18, fig.height= 7}
plot_RLE(assay(transcriptome_normalized,'RUV_K10_on_HTseq_counts')[,600:1168], transcriptome_normalized@colData$batch[600:1168], ylimit = c(-4,4))


```
Save the data

```{r}
# saveRDS(list(transcriptome_normalized, methylation_normalized), file = 'Data/TCGA_BRCA_Normalized_transcriptome&methylation.rds')

```


Explore the CPTAC proteome data

```{r}

TCGA_Breast_BI_Proteome_itraq <- read_delim("Data/TCGA_Breast_BI_Proteome.itraq.tsv", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)

library(readxl)
CPTAC_TCGA_BreastCancer_select_clinical_data_r1 <- read_excel("Data/CPTAC_TCGA_BreastCancer_select_clinical_data_r1.xlsx", 
    sheet = "Nature2012_subtype_calls")

TCGA_Breast_BI_Proteome_itraq1 <- as.data.frame(TCGA_Breast_BI_Proteome_itraq[,-1])
rownames(TCGA_Breast_BI_Proteome_itraq1) <- unlist(TCGA_Breast_BI_Proteome_itraq[,1])

data_proteome <- SummarizedExperiment(assays = list(log_ratios = TCGA_Breast_BI_Proteome_itraq1[-(1:3),-(223:228)]),
                                      colData = list(t(TCGA_Breast_BI_Proteome_itraq1[(1:3),-(223:228)])),
                                      rowData = list(TCGA_Breast_BI_Proteome_itraq1[-(1:3),(223:228)]))


# add extra metadata downloaded from the study files on PDC
extra_coldata <- as.data.frame(CPTAC_TCGA_BreastCancer_select_clinical_data_r1[-c(1:5),])
colnames(extra_coldata) <- CPTAC_TCGA_BreastCancer_select_clinical_data_r1[5,]
rownames(extra_coldata) <- substr(extra_coldata$`Complete TCGA ID`, 6,12)
extra_coldata1 <- extra_coldata[match(substr(colnames(data_proteome),1,7), substr(extra_coldata$`Complete TCGA ID`, 6,12)),]
rownames(extra_coldata1) <- colnames(data_proteome)
data_proteome@colData <- cbind(data_proteome@colData, extra_coldata1)
data_proteome <- data_proteome[, -which(is.na(data_proteome@colData$`PAM50 mRNA`))]

# delete lowly expressed proteins
na_fraction <- apply(assay(data_proteome), 1, function(x) {sum(is.na(x))/length(x)})
data_proteome <- data_proteome[which(na_fraction <= 0.3),]
assay(data_proteome) <- replace_all_NAs_with_row_means(assay(data_proteome))

# split shared & unshared into 2 assays
unshared_proteome <- data_proteome[,grepl('Unshared', colnames(data_proteome))]
data_proteome <- data_proteome[,!grepl('Unshared', colnames(data_proteome))]
colnames(unshared_proteome) <- colnames(data_proteome)
assay(data_proteome, 'unshared_log_ratio') <- assay(unshared_proteome)
assayNames(data_proteome)[1] <- 'shared_log_ratio'

saveRDS(data_proteome, file = 'Data/TCGA-BRCA CPTAC proteome.rds')

```


```{r}

plot_3D_PCA(assay(data_proteome,1), data_proteome@colData$`PAM50 mRNA`, is_PCA_obj = F)


```

```{r}
plot_3D_PCA(assay(data_proteome,2), data_proteome@colData$`PAM50 mRNA`, is_PCA_obj = F)

```


```{r, fig.width= 13, fig.height= 8}

# add purity data from rnaseq
transcriptome_normalized@colData$Purity_singscore[match(substr(transcriptome_normalized@colData$sample__RNAseq, 1,12),data_proteome@colData$`Complete TCGA ID`)]
data_proteome@colData$singscore_purity_rna <- transcriptome_normalized@colData$Purity_singscore[match(data_proteome@colData$`Complete TCGA ID`, substr(transcriptome_normalized@colData$sample__RNAseq, 1,12))]
data_proteome@colData$singscore_purity_rna[is.na(data_proteome@colData$singscore_purity_rna)] <- mean(data_proteome@colData$singscore_purity_rna, na.rm = TRUE)


# add batch data
CPTAC_TCGA_Breast_Cancer_iTRAQ_Sample_Mapping <- read_excel("Data/CPTAC_TCGA_Breast_Cancer_iTRAQ_Sample_Mapping.xlsx")
proteome_batch_data <-tidyr::pivot_longer(as.data.frame(t(apply(CPTAC_TCGA_Breast_Cancer_iTRAQ_Sample_Mapping, 1, function(x) {substr(x[3:5],6,16)}))[-(37:41),]), cols = everything())
proteome_batch_data$batch <- rep(1:36, each = 3)
proteome_batch_data <- proteome_batch_data[match(substr(data_proteome@colData@rownames,1,11) , proteome_batch_data$value),]
proteome_batch_data$batch[which(is.na(proteome_batch_data$batch))] <- 33
data_proteome@colData$batch <- proteome_batch_data$batch

```

```{r, fig.width=12, fig.height= 6}

p1 <- plot_PCA(assay(data_proteome,2),'unshared proteome log ratio', data_proteome@colData$`PAM50 mRNA`, 'pam50 (RNAseq)', is_pca_obj = F)
p2 <- plot_PCA(assay(data_proteome,2),'unshared proteome log ratio', data_proteome@colData$`PAM50 mRNA`, 'pam50 (RNAseq)', is_pca_obj = F, pcs=c(1,3))

gridExtra::grid.arrange(p1,p2, ncol=2)

```

```{r, fig.width = 12, fig.height= 6}
plot_RLE(assay(data_proteome,1), data_proteome@colData$batch, ylim = c(-4,4))

```

```{r, fig.width = 12, fig.height= 6}
plot_RLE(sweep(assay(data_proteome,1), 2, data_proteome@colData$Median, FUN = '+'), data_proteome@colData$batch, ylim = c(-4,4))


```

```{r, fig.width= 12, fig.height= 6}
plot_RLE(assay(proteome_normalized,4),proteome_normalized@colData$Median, is_continuous = T, ylimit= c(-4,4), variable_name = 'median reporter ion log ratio')
```

```{r, fig.width= 13, fig.height= 8}

assay(data_proteome, 'raw_shared_log_ratios') <- sweep(assay(data_proteome,1), 2, data_proteome@colData$Median, FUN = '+')
assay(data_proteome, 'raw_unshared_log_ratios') <- sweep(assay(data_proteome,2), 2, unshared_proteome@colData$Median, FUN = '+')

p1 <- plot_linear_correlation(assays(data_proteome), c('shared log ratio','unshared log ratio','raw_shared_log_ratios','raw_unshared_log_ratios'), data_proteome@colData$singscore_purity_rna, 'purity (rnaseq)', is.pca.obj = F)
p2 <- plot_vector_correlation(assays(data_proteome), c('shared log ratio','unshared log ratio','raw_shared_log_ratios','raw_unshared_log_ratios'), data_proteome@colData$`PAM50 mRNA`, 'pam50 subtype', is.pca.obj = F)
p3 <- plot_linear_correlation(assays(data_proteome), c('shared log ratio','unshared log ratio','raw_shared_log_ratios','raw_unshared_log_ratios'), data_proteome@colData$Median, 'log expression median', is.pca.obj = F)
p4 <- plot_vector_correlation(assays(data_proteome), c('shared log ratio','unshared log ratio','raw_shared_log_ratios','raw_unshared_log_ratios'), data_proteome@colData$batch, 'batch', is.pca.obj = F)

gridExtra::grid.arrange(p1,p2,p3,p4, ncol=2, nrow=2)

```

Since each batch is only 3 samples, we might try combining batches with similar average PC scores for the first 10 PCs

```{r}

proteome_pca_scores <- as.data.frame(run_PCA(assay(data_proteome,'raw_unshared_log_ratios'))$u)
proteome_pca_scores$batch <- data_proteome@colData$batch
avg_pc_scores <- aggregate(. ~ batch, data= proteome_pca_scores, FUN = mean)
library(dynamicTreeCut)
distances <- dist(x=avg_pc_scores[,-1])
hc <- hclust(distances)
# Use dynamic cut
dynamic_clusters <- cutreeDynamic(dendro=hc,minClusterSize=3)
avg_pc_scores$combined_batch <- dynamic_clusters


plot_ly(data = avg_pc_scores[,-c(1,12)], 
          x = ~PC1, y = ~PC2, z = ~PC3, color = avg_pc_scores$combined_batch,
          type = 'scatter3d', mode = 'markers',
          marker = list(size = 4, opacity = 0.8)) %>%
    layout(title = title,
           scene = list(xaxis = list(title = 'PC1'),
                        yaxis = list(title = 'PC2'),
                        zaxis = list(title = 'PC3')))


```


```{r}

batch_mapping <- setNames(avg_pc_scores$combined_batch, avg_pc_scores$batch)

data_proteome@colData$combined_batch <- factor(as.vector(batch_mapping[as.character(data_proteome@colData$batch)]))
data_proteome@colData$subtypes <- factor(data_proteome@colData$`PAM50 mRNA`)

plot_vector_correlation(assays(data_proteome), names(assays(data_proteome)), data_proteome$combined_batch, 'combined batch', is.pca.obj = F)


```

```{r, fig.width= 13, fig.height= 8}

proteome_normalized <- RUVIIIPRPS::normalise(data_proteome, 'raw_unshared_log_ratios', 
                                             apply.log = F,
                                             k= c(4,8,12) ,
                                             bio.variable.prps = 'subtypes',
                                             uv.variables.prps = c('Median','combined_batch', 'singscore_purity_rna'),
                                             batch.variable.prps = 'combined_batch',
                                             norm.assay.name.ncg = 'raw_unshared_log_ratios',
                                             bio.variables.ncg = 'subtypes',
                                             uv.variables.ncg = c('Median','combined_batch','singscore_purity_rna'),
                                             assess.cor.variables.prps = F,
                                             assess.se.obj = F,
                                             min.sample.for.prps = 2,
                                             min.sample.per.batch.prps = 3)

normalized_proteome_pcas <- lapply(assays(proteome_normalized), run_PCA)

p1 <- plot_linear_correlation(normalized_proteome_pcas, names(assays(proteome_normalized)), proteome_normalized@colData$singscore_purity_rna, 'purity (rnaseq)', is.pca.obj = T)
p2 <- plot_vector_correlation(normalized_proteome_pcas, names(assays(proteome_normalized)), proteome_normalized@colData$subtypes, 'pam50 subtype', is.pca.obj = T)
p3 <- plot_linear_correlation(normalized_proteome_pcas, names(assays(proteome_normalized)), proteome_normalized@colData$Median, 'median', is.pca.obj = T)
p4 <- plot_vector_correlation(normalized_proteome_pcas, names(assays(proteome_normalized)), proteome_normalized@colData$combined_batch, 'combined batch', is.pca.obj = T)

gridExtra::grid.arrange(p1,p2,p3,p4, ncol=2, nrow=2)


```

```{r, fig.height= 6}
p1 <- plot_count_map(proteome_normalized@colData$combined_batch, proteome_normalized@colData$subtypes, 1)
p2 <- plot_count_map(proteome_normalized@colData$batch, proteome_normalized@colData$subtypes, 1)

gridExtra::grid.arrange(p1,p2, nrow= 2)

```

```{r}

plot_3D_PCA(assay(proteome_normalized, 'RUV_K12_on_raw_unshared_log_ratios'), proteome_normalized@colData$subtypes, is_PCA_obj = F)

```

```{r}
# transcript_subtype_ftest <- ftest(assay(transcriptome_normalized, 'RUV_K8_on_HTseq_counts'), transcriptome_normalized@colData$subtype)
# transcript_subtype_ftest1 <- transcript_subtype_ftest[order(transcript_subtype_ftest[['Adj.PValue']]),]

```

```{r, fig.width= 12, fig.height= 6}

p1 <- plot_PCA(normalized_transcriptome_pcas[[6]], 'normalized transcriptome', as.vector(assay(transcriptome_normalized['FOXA1',],"RUV_K8_on_HTseq_counts")), 'FOXA1 expression', is_pca_obj = T, is_continuous = T)
p2 <- plot_PCA(normalized_transcriptome_pcas[[6]], 'normalized transcriptome', as.vector(assay(transcriptome_normalized['COL10A1',],"RUV_K8_on_HTseq_counts")), 'MLPH expression', is_pca_obj = T, is_continuous = T)

gridExtra::grid.arrange(p1,p2, ncol=2)

```
```{r}
cancor(colMedians(compute_log_transformed_RLE(assay(data_proteome, 'raw_unshared_log_ratios'))), fastDummies::dummy_cols(data_proteome$batch)[,-1])$cor

proteome_rle_medians <- colMedians(compute_log_transformed_RLE(assay(data_proteome, 'raw_unshared_log_ratios')))
plot_linear_correlation(assays(proteome_normalized), names(assays(proteome_normalized)), proteome_rle_medians, 'rle medians', is.pca.obj = F)

```

```{r, fig.width=12, fig.height= 6}
plot_RLE(assay(proteome_normalized, "RUV_K8_on_raw_unshared_log_ratios"), proteome_normalized@colData$batch, ylimit= c(-4,4))

```


```{r}
distances <- dist(proteome_rle_medians)
hc <- hclust(distances)
# Use dynamic cut
dynamic_clusters <- cutreeDynamic(dendro=hc,minClusterSize=4)
data_proteome@colData$combined_batch_rle_medians <- factor(dynamic_clusters)

plot_vector_correlation(assays(proteome_normalized), names(assays(proteome_normalized)), data_proteome@colData$combined_batch_rle_medians, 'rle medians batch', is.pca.obj = F)

```

```{r, fig.width= 13}

proteome_normalized_rle_medians <- RUVIIIPRPS::normalise(data_proteome, 'raw_unshared_log_ratios', 
                                             apply.log = F,
                                             k= c(4,8,12) ,
                                             bio.variable.prps = 'subtypes',
                                             uv.variables.prps = c('Median','combined_batch_rle_medians', 'singscore_purity_rna'),
                                             batch.variable.prps = 'combined_batch_rle_medians',
                                             norm.assay.name.ncg = 'raw_unshared_log_ratios',
                                             bio.variables.ncg = 'subtypes',
                                             uv.variables.ncg = c('Median','combined_batch_rle_medians','singscore_purity_rna'),
                                             assess.cor.variables.prps = F,
                                             assess.se.obj = F,
                                             min.sample.for.prps = 2,
                                             min.sample.per.batch.prps = 3)

```
```{r, fig.width = 14, fig.height=8}
normalized_proteome_pcas1 <- lapply(assays(proteome_normalized_rle_medians), run_PCA)

p1 <- plot_linear_correlation(normalized_proteome_pcas1, names(assays(proteome_normalized_rle_medians)), proteome_normalized_rle_medians@colData$singscore_purity_rna, 'purity (rnaseq)', is.pca.obj = T)
p2 <- plot_vector_correlation(normalized_proteome_pcas1, names(assays(proteome_normalized_rle_medians)), proteome_normalized_rle_medians@colData$subtypes, 'pam50 subtype', is.pca.obj = T)
p3 <- plot_linear_correlation(normalized_proteome_pcas1, names(assays(proteome_normalized_rle_medians)), proteome_normalized_rle_medians@colData$Median, 'median', is.pca.obj = T)
p4 <- plot_vector_correlation(normalized_proteome_pcas1, names(assays(proteome_normalized_rle_medians)), proteome_normalized_rle_medians@colData$combined_batch_rle_medians, 'rle_medians batch', is.pca.obj = T)

gridExtra::grid.arrange(p1,p2,p3,p4, ncol=2, nrow=2)
```


```{r, fig.width=12, fig.height= 6}

plot_RLE(assay(proteome_normalized_rle_medians,'RUV_K12_on_raw_unshared_log_ratios'), proteome_normalized_rle_medians@colData$combined_batch_rle_medians, ylimit = c(-4,4))

```

```{r, fig.width= 6, fig.height= 3}
plot_count_map(proteome_normalized_rle_medians@colData$combined_batch_rle_medians, proteome_normalized_rle_medians@colData$subtypes, color_threshold = 1) 

```

```{r}

empty <- MultiAssayExperiment()
data(miniACC)

```


Muck around with WGCNA for eigengene summarization

```{r}
sft = pickSoftThreshold(t(assay(transcriptome_normalized, 'RUV_K8_on_HTseq_counts')), powerVector = c(1:10))
plot(x=sft$fitIndices$Power, y=sft$fitIndices$SFT.R.sq)
```

```{r, fig.width=10}
adjacency <- adjacency(t(assay(transcriptome_normalized, 'RUV_K8_on_HTseq_counts'))[,1:7000], power= 3)
TOM = TOMsimilarity(adjacency)
TOMdist <- 1-TOM

geneTree = hclust(d=as.dist(1-TOM), method = "average")
dynamicMods = cutreeDynamic(dendro = geneTree, distM = (1-TOM), # deepSplit = 2, pamRespectsDendro = FALSE
                            cutHeight = 0.95)

table(dynamicMods)

plot(geneTree, xlab = '', sub= '', labels = F)
```

```{r, fig.height= 5}

module <- 2

p1 <- pheatmap(assay(transcriptome_normalized, 'RUV_K8_on_HTseq_counts')[as.vector(which(dynamicMods == module)),1:50],
         show_rownames = T, 
         show_colnames = F,
         clustering_distance_rows = "euclidean", 
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         color = colorRampPalette(c("black", 'green'))(255),
         border_color = NA,
         fontsize_row = 7, 
         )

eigengene_scores <- (run_PCA(assay(transcriptome_normalized, 'RUV_K8_on_HTseq_counts')[as.vector(which(dynamicMods == module)),1:50])$u[,1])[p1$tree_col$order]
eigengene_data <- data.frame(patient = substr(names(eigengene_scores), 9,12), score = as.vector(eigengene_scores))

ggplot(eigengene_data, aes(y=score, x=patient))+
  geom_bar(stat = "identity", 
           aes(fill = score > 0), 
           show.legend = FALSE) + 
  scale_fill_manual(values = c('red', "green")) + 
  # scale_color_manual(values = c("green", "red")) + 
  theme_minimal() + 
  labs(x = "Samples", y = "Eigengene expression", title = "Eigengene Expression")+
  theme(axis.text.x = element_text(size = 0,angle = 45,hjust = 1),
        panel.border=element_rect(colour = "grey90", fill=NA, size=0.8),
        axis.line = element_line(colour = "grey90", linewidth = 0.8),
        panel.grid.major = element_line(color = "grey98"))

```

```{r}




```



















